python/semantic_kernel/connectors/memory/sql_server.py

```diff
@@ -0,0 +1,1028 @@
+            newline: Whether to add a newline after the table name or suffix.
+        """
+        self.append(f"{prefix} [{schema}].[{table_name}] {suffix or ''}", suffix="\n" if newline else "")
+
+    def remove_last(self, number_of_chars: int):
+        """Remove the last number_of_chars from the StringBuilder."""
+        current_pos = self._file_str.tell()
+        if current_pos >= number_of_chars:
+            self._file_str.seek(current_pos - number_of_chars)
+            self._file_str.truncate()
+
+    @contextmanager
+    def in_parenthesis(self, prefix: str | None = None, suffix: str | None = None):
+        """Context manager to add parentheses around a block of strings.
+
+        Args:
+            prefix: Optional prefix to add before the opening parenthesis.
+            suffix: Optional suffix to add after the closing parenthesis.
+
+        """
+        self.append(f"{prefix or ''} (")
+        yield
+        self.append(f") {suffix or ''}")
+
+    @contextmanager
+    def in_logical_group(self):
+        """Create a logical group with BEGIN and END."""
+        self.append("BEGIN", suffix="\n")
+        yield
+        self.append("\nEND", suffix="\n")
+
+    def __str__(self):
+        """Return the string representation of the StringBuilder."""
+        return self._file_str.getvalue()
+
+
+class SqlCommand:
+    """A class that represents a SQL command with parameters."""
+
+    def __init__(
+        self,
+        query: QueryBuilder | str | None = None,
+        execute_many: bool = False,
+    ):
+        """Initialize the SqlCommand.
+
+        This only allows for creation of the query string, use the add_parameter
+        and add_parameters methods to add parameters to the command.
+
+        Args:
+            query: The SQL command string or QueryBuilder object.
+            execute_many: Whether to execute the command many times.
+                If True, the parameters will be added to each of the many_parameters contents.
+                If False, the parameters will be added to the parameters list.
+
+        """
+        self.query = QueryBuilder(query)
+        self.parameters: list[str] = []
+        self.many_parameters: list[tuple[str, ...]] = []
+        self.execute_many: bool = execute_many
+
+    def add_parameter(self, value: str) -> None:
+        """Add a parameter to the SqlCommand.
+
+        This adds a single value to the parameters.
+        If the command is set to execute many, it will add the parameters
+        to each of the the many_parameters contents.
+        Or create a new tuple with the value if there are no many_parameters yet.
+        """
+        if self.execute_many:
+            if not self.many_parameters:
+                self.many_parameters.append((value,))
+                return
+            for i in range(len(self.many_parameters)):
+                self.many_parameters[i] += (value,)
+        else:
+            if (len(self.parameters) + 1) > 2100:
```

Comment:
For very large upserts, we can still easily exceed 2100 if each record has many fields. Do we need more sophisticated batching logic?

Reply:
in a follow-up, yes